### PR TITLE
Add option to publish the sbpl footprint plan

### PR DESCRIPTION
**Pull request by LotfiZ**
Hello folks,

I added the methods already defined in the sbpl_cart_planner to display the plan with the footprint, it can be useful for debugging or tuning purposes. i dont know if its relevant or not ! 

Cheers

### CODE REVIEW

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 1]
```diff
@@ -206,6 +210,7 @@ void SBPLLatticePlanner::initialize(std::string name, costmap_2d::Costmap2DROS*
     ROS_INFO("[sbpl_lattice_planner] Initialized successfully");
     plan_pub_ = private_nh.advertise<nav_msgs::Path>("plan", 1);
     stats_publisher_ = private_nh.advertise<sbpl_lattice_planner::SBPLLatticePlannerStats>("sbpl_lattice_planner_stats", 1);
+    sbpl_plan_footprint_pub_ = private_nh.advertise<visualization_msgs::Marker>("sbpl_plan_footprint_array", 1);
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp]: Maybe a shorter name, like "footprint_markers", would be better? It's in the private namespace anyway, so there won't be any name collisions. Also, it's not an array.

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 2]
```diff
@@ -460,4 +470,69 @@ bool SBPLLatticePlanner::makePlan(const geometry_msgs::PoseStamped& start,
...
+                                        visualization_msgs::Marker & ma)
+ { 
+   //ma.resize(1); 
+   ma.header.frame_id = path_frame_id;
+   ma.header.stamp = ros::Time();
+   ma.ns = "sbpl_robot_footprint";
+   ma.id = 0;
+   ma.type = visualization_msgs::Marker::LINE_LIST;
+   ma.action = visualization_msgs::Marker::ADD;
+   ma.lifetime = ros::Duration(2);
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp]: 2 seconds is a bit short in my opinion. I'd suggest setting the lifetime to "0". That way, the marker stays visible until the next marker is published.

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 3]
```diff
@@ -460,4 +470,69 @@ bool SBPLLatticePlanner::makePlan(const geometry_msgs::PoseStamped& start,
...
+   ma.scale.x = 0.05;
+   ma.color.a = 1.0;
+   ma.color.r = 0.0;
+   ma.color.g = 0.0;
+   ma.color.b = 1.0;
+
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp]: RViz shows the following warning: "Uninitialized quaternion, assuming identity."

To fix it, set `ma.pose.orientation.w = 1.0`.

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 4]
```diff
@@ -460,4 +470,69 @@ bool SBPLLatticePlanner::makePlan(const geometry_msgs::PoseStamped& start,
...
+     robot_pose.position.z = 0.0;
+     tf::Quaternion quat;
+     quat.setRPY(0.0,0.0,sbpl_path[i].theta);
+     tf::quaternionTFToMsg(quat,robot_pose.orientation);
+     transformFootprintToEdges(robot_pose,robot_footprint_,transformed_rfp);
+ 
+     for(unsigned int i=0; i < transformed_rfp.size(); i++)
+       ma.points.push_back(transformed_rfp[i]);
+ 
+
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp]: Please remove these 2 empty lines.

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 5]
```diff
@@ -460,4 +470,69 @@ bool SBPLLatticePlanner::makePlan(const geometry_msgs::PoseStamped& start,
...
+     {
+       out_footprint[2*i-1].x = out_footprint[2*i].x;
+       out_footprint[2*i-1].y = out_footprint[2*i].y;
+     }
+   }
+   return;
+ }
+
+
+
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp:510]: Please remove these 3 empty lines.

**Review by mintar — CHANGES_REQUESTED** [review]
Thanks for your contribution, this is definitely useful! I've added some requested changes in the comments.

**Comment by LotfiZ** [comment]
I did the requested changes, if i forgot something i'll happy to fix it. Thank you

File: sbpl_lattice_planner/include/sbpl_lattice_planner/sbpl_lattice_planner.h [hunk 6]
```diff
@@ -92,11 +102,13 @@ class SBPLLatticePlanner : public nav_core::BaseGlobalPlanner{
   std::string name_;
   costmap_2d::Costmap2DROS* costmap_ros_; /**< manages the cost map for us */
   std::vector<geometry_msgs::Point> footprint_;
+  std::vector<geometry_msgs::Point> robot_footprint_;
```
Comment (mintar) [inline sbpl_lattice_planner/include/sbpl_lattice_planner/sbpl_lattice_planner.h]: Why is the extra copy `robot_footprint_` necessary? If I see it correctly, we could just re-use `footprint_`, right?

File: sbpl_lattice_planner/src/sbpl_lattice_planner.cpp [hunk 7]
```diff
@@ -42,6 +42,7 @@
 
 #include <costmap_2d/inflation_layer.h>
 #include <tf2/LinearMath/Quaternion.h>
+#include <tf/transform_broadcaster.h>
```
Comment (mintar) [inline sbpl_lattice_planner/src/sbpl_lattice_planner.cpp]: I don't think we need to include the `transform_broadcaster` here. Maybe some more specific includes though? I haven't checked.

**Comment by mintar** [comment]
Thanks for your changes! I only have two more comments before this can be merged (see above).

**Comment by LotfiZ** [comment]
Hi Martin, sorry for the late, i made the requested changes but i'm not sure about the TF header that i included, if there is a more optimal way, i can change it ! Thank you

**Review by mintar — APPROVED** [review]

**Comment by mintar** [comment]
Thanks! I've added a few cleanups and merged the PR.

**Pull request merged**